### ISSUE TITLE
GCSToGCSOperator delimiter bug

### ISSUE THREAD

**Issue by yukijam** [apache/airflow#29115]
### Apache Airflow version

2.5.1

### What happened

Using GCSToGCSOperator with delimiter = '.json' would also consider the incorrect files
for example: I got a file named xxxx.jsonl__ 

### What you think should happen instead

it should exclude the files not ending with .json

### Operating System

gcp airflow composer

### Deployment

Composer

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
assigned you @yukijam  if you want to fix it

**Comment by Taragolis** [comment]
Due to name of argument as `delimiter`, description and code of Operator and [google-cloud-storage](https://cloud.google.com/python/docs/reference/storage/latest/buckets#listblobsmaxresultsnone-pagetokennone-prefixnone-delimiternone-startoffsetnone-endoffsetnone-includetrailingdelimiternone-versionsnone-projectionnoacl-fieldsnone-clientnone-timeout60-retrygoogleapicoreretryretry-object) I thought it work correctly. However have no idea who actually need this functionality.

I thought better implement new argument with proper implementation and with name `suffix`

**Comment by shahar1** [comment]
@eladkal Please assign me

**Comment by potiuk** [comment]
Done

**Comment by shahar1** [comment]
> Due to name of argument as `delimiter`, description and code of Operator and [google-cloud-storage](https://cloud.google.com/python/docs/reference/storage/latest/buckets#listblobsmaxresultsnone-pagetokennone-prefixnone-delimiternone-startoffsetnone-endoffsetnone-includetrailingdelimiternone-versionsnone-projectionnoacl-fieldsnone-clientnone-timeout60-retrygoogleapicoreretryretry-object) I thought it work correctly. However have no idea who actually need this functionality.
> 
> I thought better implement new argument with proper implementation and with name `suffix`

Based on the following resources:
[1] https://cloud.google.com/storage/docs/json_api/v1/objects/list?apix_params=%7B%22bucket%22%3A%22reproducing_issue_11874%22%2C%22delimiter%22%3A%22.PNG_2%22%7D#try-it
[2] https://stackoverflow.com/questions/44702835/google-cloud-storage-ignoring-delimiter


It seems that treating `delimiter` as `suffix` is a misuse, and filtering according to suffixes shouldn't be done as part of the API call to GCS but as a separate logic within the hook by simple list filtering.
Therefore, I suggest deprecating the `delimiter` parameter (as it should almost always be `"/") and adding a `suffix` parameter instead for use as a filter, as follows:
If `delimiter` is set and `suffix` is not already provided, warn that `delimiter` is deprecated and assign `self.suffix` its value, while `self.delimiter` will be set to a constant `"/"`. If `suffix` is provided, it will behave as expected regardless of the value of `delimiter`.

**Comment by shahar1** [comment]
> Due to name of argument as `delimiter`, description and code of Operator and [google-cloud-storage](https://cloud.google.com/python/docs/reference/storage/latest/buckets#listblobsmaxresultsnone-pagetokennone-prefixnone-delimiternone-startoffsetnone-endoffsetnone-includetrailingdelimiternone-versionsnone-projectionnoacl-fieldsnone-clientnone-timeout60-retrygoogleapicoreretryretry-object) I thought it work correctly. However have no idea who actually need this functionality.
> 
> I thought better implement new argument with proper implementation and with name `suffix`

Based on the following resources:
[1] https://cloud.google.com/storage/docs/json_api/v1/objects/list?apix_params=%7B%22bucket%22%3A%22reproducing_issue_11874%22%2C%22delimiter%22%3A%22.PNG_2%22%7D#try-it
[2] https://stackoverflow.com/questions/44702835/google-cloud-storage-ignoring-delimiter


It seems that treating `delimiter` as `suffix` is a misuse, and filtering according to suffixes shouldn't be done as part of the API call to GCS but as a separate logic within the hook by simple list filtering.
Therefore, I suggest deprecating the `delimiter` parameter (as it should almost always be `"/") and adding a `suffix` parameter instead for use as a filter, as follows:
If `delimiter` is set and `suffix` is not already provided, warn that this parameter is deprecated and assign `self. Suffix` its value (), while `self.delimiter` will be set to `"/"`. This usage of the `delimiter` parameter is incorrect, but it won't break if already used in this context.
2. If `delimiter` is set, warn that this parameter is deprecated and completely ignore its value. This may break existing usages of `delimiter` in this context.

**Comment by eladkal** [comment]
> filtering according to suffixes shouldn't be done as part of the API call to GCS but as separate logic within the hook

I have ambivalent feelings about hooks doing more than the API/service allows.
If GCS decided to offer `prefix` & `delimiter` I'm not sure if we should create a new functionally?

I think it would be easier to discuss over a PR where we can also see the test cases and get a sense of how much further customizations we do from Airflow side.

**Comment by shahar1** [comment]
> > filtering according to suffixes shouldn't be done as part of the API call to GCS but as separate logic within the hook
> 
> I have ambivalent feelings about hooks doing more than the API/service allows. If GCS decided to offer `prefix` & `delimiter` I'm not sure if we should create a additional functionally?
> 
> I think it would be easier to discuss over a PR where we can also see the test cases and get a sense of how much further customizations we do from Airflow side.

I created a PR draft (#31107) - the bottom line is that the current implementation is suboptimal, and a more thorough change should be done. There are two main issues regarding the GCS hook/operators:
1. As I previously stated, `delimiter` is mostly misused as `suffix`. The right way to do this filtering is by utilizing a relatively new parameter called `matchGlob`, which utilizes glob patterns to filter objects based on their paths. Unfortunately, the GCS Python client doesn't support `matchGlob` yet - but it's quite easy to patch the appropriate function until it is natively integrated.

2. Another issue in the same area is that we have a *very* complex logic for supporting wildcards (*) in the source object names, which is not supported by the GCS client at all. I'd like to deprecate the usage of wildcards as it would become very difficult to handle along with the deprecation of the `delimiter`. Should I treat it in a different PR and/or create a new issue for that? (I implemented the solution in the draft, but I'm afraid that it might make things complex to test)
